### PR TITLE
Access control circumvention fix

### LINKED ISSUE #743: Access Control Easy to Circumvent
The bot clearly states that the assignee does not have access to set the `Priority` label, thus funding the bounty. However, it appears that they were able to edit an invalid label that was already on the issue, and rename it to a valid label to generate a payment permit.  

## Proposed Solution

- If anybody but an `admin` or `billing_manager` updates pricing related labels, then an `admin` or `billing_manager` must manually re-allow automatic payment permit generation. This state should be saved in our database. 

- The bot will check the last updated time of a dependent (pricing related) label and if it was between the time the issue was started and completed. This implies that one of the dependent labels was tampered with while a task was being worked on. 

### Example:

- Issue has `Priority: 4 (Urgent)` and `Time: <1 Hour`
- In our database we have the `updated` column for `Priority: 4 (Urgent)`
- The issue was started on by an assignee on 1 October. 
- The `Priority: 4 (Urgent)` label was last modified, by an unauthorized contributor, on 2 October. 
- The issue is marked as complete.
- Due to the dependent label `Priority: 4 (Urgent)` being modified after the task was started, the bot refuses to generate the permit and explains the situation.

### Notes

- this should only be concerned with pricing related labels defined in our config 
- In this scenario, if it is still not authorized, then the `admin` or `billing_manager` should change the label back from Y to X and the bot should re-allow permits to be generated. 
- when it is authorized, perhaps we should make a command like `/approve` for `admins`, and `billing_managers` which will "approve" in the database see below.
- Our access control should accomodate for these as well. For example if I wrote `/allow set-priority @seprintour true` and if seprintour changed the priority label, it should be approved implicitly. 

### Unauthorized label changes table
1. id, int8
2. created, timestamptz
3. updated, timestamptz
4. user id, int8
5. repository id, int8
6. label from, string
7. label to, string
8. authorized, bool

> @seprintour, You are not allowed to add Priority: 3 (High)

_Originally posted by @ubiquibot[bot] in https://github.com/ubiquity/telegram-ubiquibot/issues/4#issuecomment-1710631705_
            

### Task checklist

1. create "label changes" table
2. intercept label change event and log if they dont have appropriate permissions
3. create `/approve` command (i feel like this part of the proposal can be refined... rfc!)

### PR DESCRIPTION
**Pull request by seprintour**
Resolves #743 

Quality Assurance: https://github.com/Seprintour-Test/test/issues/44

### CODE REVIEW

File: app.yml
Comment (0x4007) [inline app.yml:1]: @0xcodercrane is this file still needed or is it an artifact from when this used to run entirely as a GitHub action?

Comment (seprintour) [reply app.yml:1]: > @0xcodercrane is this file still needed or is it an artifact from when this used to run entirely as a GitHub action?

It's still needed, even as a GitHub app.. It controls the permission GitHub grants the bot, that's why i uncommented `label`.

So once the bot is installed or updated, it asks for permission to listen to that event

File: src/handlers/payout/action.ts [hunk 1]
```diff
@@ -194,6 +194,17 @@ export const incentivesCalculation = async (): Promise<IncentivesCalculationResu
...
   }
 
+  // check for label altering here
+  const labelChanges = await getLabelChanges(repository.full_name, [issueDetailed.priceLabel, issueDetailed.priorityLabel, issueDetailed.timelabel]);
+
+  if (labelChanges) {
+    // if approved is still false, it means user was certainly not authorized for that edit
+    if (!labelChanges.approved) {
+      logger.info(`Skipping... label was changed by unauthorized user`);
+      throw new Error(`Permit generation disabled because this label: ${labelChanges.label_to} on this issue was edited by an authorized user`);
```
Comment (0x4007) [inline src/handlers/payout/action.ts]:
```suggestion
      throw new Error(`Permit generation disabled because label "${labelChanges.label_to}" was modified by an unauthorized user`);
```

File: supabase/migrations/20230924193504_label_changes.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS label_changes (
+    id serial PRIMARY KEY,
+    created_at timestamptz NOT NULL,
+    updated_at timestamptz NOT NULL,
+    username text NOT NULL,
+    repository text NOT NULL,
+    label_from text NOT NULL,
+    label_to text NOT NULL,
+    approved boolean NOT NULL
```
Comment (0x4007) [inline supabase/migrations/20230924193504_label_changes.sql]: I think we will also need a command to approve.

Comment (seprintour) [reply supabase/migrations/20230924193504_label_changes.sql]: Yea, i am working on that now

File: src/handlers/comment/handlers/approve.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    logger.info(`Skipping... its not a bounty`);
+    return ErrorDiff(`No valid bounty label on this issue`);
+  }
+
+  // check for label altering here
+  const labelChanges = await getLabelChanges(repository.full_name, [issueDetailed.priceLabel, issueDetailed.priorityLabel, issueDetailed.timelabel]);
+
+  await _approveLabelChange(labelChanges.id);
+
+  return `Label change have been approved, permit can now be generated`;
```
Comment (0x4007) [inline src/handlers/comment/handlers/approve.ts]:
```suggestion
  return `Label change has been approved, permit can now be generated`;
```

File: src/handlers/comment/handlers/index.ts [hunk 4]
```diff
@@ -304,6 +306,12 @@ export const userCommands = (): UserCommands[] => {
       handler: setAccess,
       callback: commandCallback,
     },
+    {
+      id: IssueCommentCommands.APPROVE,
+      description: `Approve a label change. (Admin Only)`,
```
Comment (0x4007) [inline src/handlers/comment/handlers/index.ts]:
```suggestion
      description: `Approve a label change. Superuser only.`,
```

File: supabase/migrations/20230924193504_label_changes.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS label_changes (
+    id serial PRIMARY KEY,
+    created timestamptz NOT NULL,
+    updated timestamptz NOT NULL,
+    username text NOT NULL,
+    repository text NOT NULL,
+    label_from text NOT NULL,
+    label_to text NOT NULL,
+    approved boolean NOT NULL
```
Comment (0x4007) [inline supabase/migrations/20230924193504_label_changes.sql]:
```suggestion
    authorized boolean NOT NULL
```
@seprintour lets change everything to `authorized` instead of `approved` and `authorize` instead of `approve`. It's important to use unique terms. Approved can refer to pull request related matters.

Comment (seprintour) [reply supabase/migrations/20230924193504_label_changes.sql]: Makes sense, i've edited that

**Review by 0x4007 — APPROVED** [review]

File: supabase/migrations/20230924193752_label_changes_rls.sql [hunk 6]
```diff
@@ -0,0 +1 @@
+ALTER TABLE label_changes ENABLE ROW LEVEL SECURITY
```
Comment (0x4007) [inline supabase/migrations/20230924193752_label_changes_rls.sql:1]: Whats this for?

Comment (seprintour) [reply supabase/migrations/20230924193752_label_changes_rls.sql:1]: > Whats this for?

To enable RLS for the DB, more like extra security..
Always enabled on all new tables

Comment (seprintour) [reply supabase/migrations/20230924193752_label_changes_rls.sql:1]: Screenshot 2023-09-26 at 12 19 31 AM

Comment (0x4007) [reply supabase/migrations/20230924193752_label_changes_rls.sql:1]: Right but why are we only now adding this?

Comment (seprintour) [reply supabase/migrations/20230924193752_label_changes_rls.sql:1]: It was added to all tables a while ago, this is just me enabling it because I created a new table
